crates/typed-store-macros/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+/// // This will fail to compile with error `All struct members must be of type DMBap<K, V>`
+/// // #[derive(DBMapUtils)]
+/// // struct BadTables {
+/// //     table1: DBMap<String, String>,
+/// //     bad_field: u32,
+/// // #}
+/// ```
+#[proc_macro_derive(DBMapUtils, attributes(options))]
+pub fn derive_dbmap_utils(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as ItemStruct);
+    let name = &input.ident;
+
+    // Check that the type is DBMap
+    let info = input.fields.iter().map(|f| {
+        // if f.attrs.len() > 1 {
+        //     panic!("Too many attributes. Only `{DB_OPTIONS_ATTR_NAME}` allowed");
+        // }
+        let attrs: Vec<_> = f
+            .attrs
+            .iter()
+            .filter(|a| a.path.is_ident(DB_OPTIONS_ATTR_NAME))
+            .collect();
+
+        let (point_lookup, cache_capacity) = if f.attrs.is_empty() {
+            (false, DEFAULT_CACHE_CAPACITY)
+        } else {
+            get_opts(attrs.get(0).unwrap()).unwrap()
+        };
+
+        let ty = &f.ty;
+        if let Type::Path(p) = ty {
+            let type_info = &p.path.segments.first().unwrap();
+            let inner_type =
+                if let PathArguments::AngleBracketed(angle_bracket_type) = &type_info.arguments {
+                    angle_bracket_type
+                } else {
+                    panic!("All struct members must be of type DMBap<K, V>");
+                };
+
+            let type_str = format!("{}", &type_info.ident);
+            // Rough way to check that this is DBMap
+            if type_str == "DBMap" {
+                return (
+                    (f.ident.as_ref().unwrap(), &f.ty),
+                    (inner_type, (point_lookup, cache_capacity)),
+                );
+            } else {
+                panic!("All struct members must be of type DMBap<K, V>");
+            }
+        }
+        panic!("All struct members must be of type DMBap<K, V>");
+    });
+
+    let (field_info, inner_types_with_opts): (Vec<_>, Vec<_>) = info.unzip();
+    let (field_names, _field_types): (Vec<_>, Vec<_>) = field_info.into_iter().unzip();
+    let (inner_types, optimizations): (Vec<_>, Vec<_>) = inner_types_with_opts.into_iter().unzip();
+    let (point_lookup, cache_capacity): (Vec<_>, Vec<_>) = optimizations.into_iter().unzip();
+
+    let precondition_str = "#[pre(\"Must be called only after `open_tables_read_only`\")]";
+    let _precondition_str_tok: proc_macro2::TokenStream = precondition_str.parse().unwrap();
+
+    TokenStream::from(quote! {
+        use std::path::PathBuf;
+        use rocksdb::Options as RocksDBOptions;
+        use anyhow::anyhow;
+        use std::collections::BTreeMap;
+        use typed_store::rocks;
+        use typed_store::reopen;
+        use typed_store::Map;
+        use rocksdb::MultiThreaded;
+        use std::path::Path as FilePath;
+        use pre::pre;
+
+        const DB_DEFAULT_CF_NAME: &str = "default";
```

Comment:
This might cause problem as well if this macro is used more then once in the same file